File: 16-trees2/1-intro/main.java
// Basic BST Node - Java Example

class BSTNode {
    int data;
    BSTNode left, right;
    BSTNode(int data) {
        this.data = data;
        left = right = null;
    }
}

public class BSTIntro {
    public static BSTNode insert(BSTNode root, int key) {
        if (root == null) return new BSTNode(key);
        if (key < root.data)
            root.left = insert(root.left, key);
        else
            root.right = insert(root.right, key);
        return root;
    }

    public static void main(String[] args) {
        BSTNode root = null;
        int[] vals = {10, 5, 15, 3, 7, 13, 17};
        for (int val : vals) {
            root = insert(root, val);
        }
        System.out.println("BST constructed with root: " + root.data);
    }
}

File: 16-trees2/1-intro/main.cpp
// Basic BST Node - C++ Example

#include <iostream>

struct BSTNode {
    int data;
    BSTNode* left;
    BSTNode* right;
    BSTNode(int val) : data(val), left(nullptr), right(nullptr) {}
};

BSTNode* insert(BSTNode* root, int key) {
    if (!root) return new BSTNode(key);
    if (key < root->data)
        root->left = insert(root->left, key);
    else
        root->right = insert(root->right, key);
    return root;
}

int main() {
    BSTNode* root = nullptr;
    int vals[] = {10,5,15,3,7,13,17};
    for (int val : vals) {
        root = insert(root, val);
    }
    std::cout << "BST constructed with root: " << root->data << std::endl;
    return 0;
}
